PR for NumberBox (#29)
microsoft/microsoft-ui-xaml-specs: active/NumberBox/NumberBox.md

<!-- This section and the appendix are the only sections that likely
do not get copied to docs.microsoft.com; they're just an aid to reading this spec. -->

<!-- If you're modifying an existing API, included a link here to the
existing page(s) -->

<!-- For example, this section is a place to explain why you're adding this API rather than
modifying an existing API. -->

<!-- For example, this is a place to provide a brief explanation of some dependent
area, just explanation enough to understand this new API, rather than telling
the reader "go read 100 pages of background information posted at ...". -->

## Description
<!-- Use this section to provide a brief description of the feature.
For an example, see the introduction to the PasswordBox control 
(http://docs.microsoft.com/windows/uwp/design/controls-and-patterns/password-box). -->

Number box is a numerics only text control with support for input formatting, validation, basic calculation, and increment stepping.

**Important APIs:** [NumberBox class](https://docs.microsoft.com/en-us/uwp/api/microsoft.ui.xaml.controls.numberbox) [TODO - generated with MSDN documentation on publication.]

## Is this the right control? 

> Use a **NumberBox** control to receive and display formatted numeric values such as currency, IP addresses, and telephone numbers. A **NumberBox** can also be enabled with basic calculator support to compute multiplication, division, addition, and subtraction across parenthetical order with standard operator precedence.

Reviewer: This still needs updating to reflect the change in scope to not include IP Address, telephone numbers, etc.

Author: Correct. I ended the 9-5 having updated the IDL the next draft of APIs and removing vestigial features from the listed examples to avoid any further community investment going into solved questions. This spec received more attention than I think any of us here in Redmond anticipated (see @jevansaks tweets) so my goal tomorrow (today?) is to revise the rest of the spec and its examples to reflect the progress we've made and create some more architect examples. 

Thanks for your patience while I am catching up on all of this **incredible community ambition!**

Author: On the note of Twitter, @mrlacey, I will accept your follow, I am just working out a bug with Twitter's help desk as I try and transfer my deprecated accounts username over to my new one. You'll hear from me soon!

> Use a **NumberBox** control to receive and display formatted numeric values such as currency, IP addresses, and telephone numbers. A **NumberBox** can also be enabled with basic calculator support to compute multiplication, division, addition, and subtraction across parenthetical order with standard operator precedence.

Reviewer 2: >addresses [](start = 97, length = 9)

how to show the validation error? For example, customer input following text for calculation?
"(1+)3+2)"
or 
"555" for part1 of an ip address.

Author: Hi, @licanhua! We have scoped out numerical strings (such as IP addresses) for NumberBox to prioritize math and computation. However , you're right, we still need to figure out exactly how to surface a validation error. I'll be bugging @LucasHaines to make sure we align NumberBox's validation behaviors with our broader validation feature that is in-progress.

## Examples
<!-- Use this section to explain the features of the API, showing
example code with each description. The general format is: 
  feature explanation,
  example code
  feature explanation,
  example code
  etc.-->
  
<!-- Code samples should be in C# and/or C++/WinRT -->

<!-- As an example of this section, see the Examples section for the PasswordBox control 
(https://docs.microsoft.com/windows/uwp/design/controls-and-patterns/password-box#examples). -->

### Create a NumberBox

XAML
```XAML
<NumberBox AutomationProperties.Name="simple NumberBox"/>
```

![](images/numberbox-basic.png)

### Add placeholder text

XAML
```XAML
<NumberBox AutomationProperties.Name="simple NumberBox" Header="Enter equation:" PlaceholderText="A + B * C" />
```

![](images/numberbox-placeholder-text.png)

### Enable calculation support

XAML
```XAML
<NumberBox AutomationProperties.Name="NumberBox for simple calculations" 

>     Header="Enter equation:" 

Reviewer 2: Like Calculate, text is aligned to right side. I prefer to set alignment to right if AcceptsCalculation is true.
Also can we preview the output. For example, if you type '1+2', there is gray text followed with '=3'

Reviewer: I spent a lot of time thinking about whether content inside the NumberBox should be aligned differently to a TextBox, as I nearly proposed this myself,  but came to the conclusion that it shouldn't. I considered what varying needs of different apps are and what Excel does (it switches internal alignment based on what it infers the content type of a cell to be.) 

I keep going back and forth on this but without hard data (I'm not sure this is even possible to obtain) for making a choice one way or another there will always be good arguments for both options. I'm also aware preference will be influenced by past personal experiences and the current needs of any developer making the decision.

My inclination is to keep the `HorizontalContentAlignment` the same as the parent control but consider changing it if during the preview the feedback from real world use of the control is that developers have to change it more than they don't.

Author: I like the idea of calculation preview assuming it doesn't introduce any unacceptable overhead! It could also be used to pre-validate correct entry for the user. These are things to consider. As far as text alignment goes, @mrlacey, we will have the pre-release period to try it out and collect feedback. My inclination would be the same as yours which is to keep HorizontalContentAlignment the same as the parent control unless we receive feedback otherwise.

Reviewer 3: A calculation preview could be displayed as a tooltip that appears above the text field perhaps?

![image](https://user-images.githubusercontent.com/7389110/58842381-cff0e280-8665-11e9-8d80-8fd509e7ce75.png)

Author: These ideas are so cool. I'm going to add an open question bullet here and on the proposal for this "preview calculations" concept.

Reviewer 3: Regarding the greyed out text example - I worry that this may encourage the user to fill in the values that appear, rather than act as a hint as to what the final result will be.  It looks just like the PlaceholderText.

What about using the AccentColor and different font weight to make it stand out
![image](https://user-images.githubusercontent.com/7389110/58920708-5b807700-872b-11e9-9924-21a7b7d37e68.png)

Author: I have the image updated here and on the Proposal. I think you are right that grayed out text could look like a prompt, which if followed, would result in an error due to the "=".

Reviewer 4: Definitely prefer the in-box preview placement compared to the thhe tooltip placement. Since the actual calculation enetred so far and the preview result are directly related to each other they also have to be visually as close to each other as possible. In the above tooltip version, that's not given (Entering calculations on the left, then have to look above the input field all to the right.) In-box placement looks about right; have some padding betwenn the input and the review value as shown above and we should be all set.

If result preview is to be implemented, we should also think about adding a new property then, like
`public bool ShowResultPreview { get; set; }`
so it can be turned on/off by users.

    MaxValue="99" />
```

## Remarks
<!-- Explanation and guidance that doesn't fit into the Examples section. -->

<!-- APIs should only throw exceptions in exceptional conditions; basically,
only when there's a bug in the caller, such as argument exception.  But if for some
reason it's necessary for a caller to catch an exception from an API, call that
out with an explanation either here or in the Examples -->

## API Notes
<!-- Option 1: Give a one or two line description of each API (type
and member), or at least the ones that aren't obvious
from their name.  These descriptions are what show up
in IntelliSense. For properties, specify the default value of the property if it
isn't the type's default (for example an int-typed property that doesn't default to zero.) -->

<!-- Option 2: Put these descriptions in the below API Details section,
with a "///" comment above the member or type. -->

## API Details
<!-- The exact API, in MIDL3 format (https://docs.microsoft.com/en-us/uwp/midl-3/) -->

```c++ 
unsealed runtimeclass NumberBox : Windows.UI.Xaml.Controls.TextBox
{
    NumberBox();
    
    Double Value;
    
    Boolean AcceptsCalculation;
    Boolean UpDownButtonsEnabled;

>     Boolean HyperDragEnabled;

Reviewer 5: What are these hyper properties?

Author: Drag **and** scroll to step the same way up/down keyboard arrows or up/down buttons would. Adobe and Blender have set a delightful precedent that our community has advocated for on the proposal. I'll give @ryandemopoulos credit for the name.

Reviewer 6: @SavoySchuler, I'm happy to help you work through the design details of all this dragging and scrolling. I think we'll need to add more detail about this behavior to the spec.

Reviewer 7: A few details based on my implementation
1. Because dragging on a `TextBox` selects text in one - in [my implementation](https://github.com/xyzzer/WinRTXamlToolkit/blob/master/WinRTXamlToolkit/Controls/NumericUpDown/NumericUpDown.cs) I chose to use a transparent but hit-testable drag overlay element that captures manipulation events when the control doesn't have focus.
2. Tapping the drag overlay makes the control (the TextBox in its template) get focus
3. As long as the `TextBox` in the template has focus - the drag overlay hit test visibility is off.
4. The mouse cursor is hidden while dragging (`Window.Current.CoreWindow.PointerCursor = null;`), which enables continuous dragging, unbounded by screen borders
5. You might want to expose a property like `DragSpeed` since `StepSize` might not necessarily match how you'd want drag input to behave.
6. You might want your dragging speed to depend on screen resolution. Not sure why this worked for me better 6 years ago. Perhaps something changed - maybe this worked for quicker hyperdragging with touch on a smaller-screen Windows phone. Perhaps on a Surface Studio or Hub - this would actually be more painful today. Perhaps calculating physical offsets based on screen DPI would make more sense today.

I have an `IsReadOnly` property that you can use to block value updates using any means. If you add that - you need to make sure the `TextBox` has a matching value, but also that up/down keys/buttons and hyperdrag don't update the value.

Author: Tagging @KevinTXY for implementation details. 

With regard to `DragSpeed` I envisioned this not as a 1:1 to `StepSize`, but rather exponentially increasing over distance (and relative to Maximum value/screen size if a Maximum value is given) - however, my priority is to align with WinRT XAML Toolkit's NumericUpDown & Blender if able. Ultimately my goal is to avoid adding an implementation requirement for the developer, particularly in the case where drag speed may not be a static value. 

 @xyzzer, with this in mind, is this still a value you believe should have writable property exposed?

Reviewer 8: HyperDrag hasn't been written and is somewhat an afterthought for me just because of it's complexity - I think I actually need to see how this works on other products first because I have doubts about it in general. By dragging are we emulating slider behavior or giving the box's value some kind of stepping velocity? How does this behave with decimal precision? Who is this for?

    Boolean HyperScrollEnabled;
    

>     Double StepSize;

Reviewer: Note that this doc must include rules to verify the StepSize relative to MaxValue and MinValue and what to do when excpetional values are provided.

Author: Agreed. I will leave this comment unresolved as a forum to discuss preference.

Author: @mrlacey, I have added this to the Behavioral Notes section with this commit: https://github.com/microsoft/microsoft-ui-xaml-specs/commit/164f9c67bcf6711606df6bc453c7da34bc16c116

  
    Boolean AreLeadingZerosTrimmed;
    Double DecimalPrecision;

>     Boolean DoesInputRound;

Reviewer: Note that this doc must specify the rounding rule(s) to use.

Author: Agreed. @jevansaks mentioned that .NET has solved this before. I believe it would be wise to align with their solution for sake of consistency across the Microsoft development space and to avoid isolated/inconsistent redundancy. I will find out what that is and add it here.

Reviewer 9: The thing I was thinking of was the WinRT [DecimalFormatter](https://docs.microsoft.com/en-us/uwp/api/windows.globalization.numberformatting.decimalformatter) and [NumberRounder](https://docs.microsoft.com/en-us/uwp/api/windows.globalization.numberformatting.significantdigitsnumberrounder) classes which we should make sure whatever properties we expose line up nicely with those types.

Reviewer 10: Should DecimalPrecision be a Double or an Integer?

Reviewer 10: I agree with @jevansaks, a DecimalFormatter & NumberRounder solution (somehow) is needed for Localization. 

Currency support using a Prefix or Suffix would be really nice.

Author: @sonnemaf and @mrlacey, agreed. There is an IncrementNumberRounder and a SignificantDigitsNumberRounder that could each be relevant to scenarios where a NumberBox is used. As of now, I have exposed the the properties necessary to set for NumberBox to internally configure your preferred rounder (specified via enum). 

we have exposed the necessary properties to allow to ease configuration of an 

@mdtauk was kind enough to open a [proposal for adding Prefix/Suffix properties to TextBox](https://github.com/microsoft/microsoft-ui-xaml/issues/784)!

    
    Double MinValue;
    Double MaxValue;
    

>     static Windows.UI.Xaml.DependencyProperty ValueProperty{ get; };

Reviewer: By exposing this ValueProperty, what does this mean for the `Text` property inherited from the `TextBox`? I assume both will still be available and ValueProperty will just be a typed version of the Text.

Author: That is correct. Additionally, our intention is to keep this in sync by having them inform each other so that changes to one will result in an automatic update to the other. The purpose of this is to have the String and Double version of the input immediately and conveniently.

Reviewer 5: What will the Value be if Text isn't a number?

Author: We'll be leaning on validation to surface error states/messages for non-numerical values and non-legal equations. I'll be speaking with @LucasHaines to align with the broader validation development.

Reviewer 3: Text is the thing you enter and see displayed.

Value is the actually computed value outputted.

The two are kept in sync

Author: Exactly, @mdtauk.

Reviewer 5: Text, being a settable property, could be set to a non-number. We should just specify the behavior. Presumably either throw or coerce Value to zero, I lean towards throwing.

Reviewer 3: @MikeHillberg I would suspect validation could be useful with this, when it comes to TextBox

Reviewer: Will requesting `Value` forcibly trigger calculation if it hasn't yet been done? I'd hope it would.
If querying `Value` and it's in an error state (of whatever kind-such as invalid calculation entered) I'd expect it to return `Double.NaN` rather than `null` as null can have meaning.

Author: @mrlacey, great feedback. I have committed this to the spec here: https://github.com/microsoft/microsoft-ui-xaml-specs/commit/8ff8217b8391516a9d013b45cf469b8ca7b3dcac

>     static Windows.UI.Xaml.DependencyProperty ValueProperty{ get; };

Reviewer: Can I correctly assume that the way this is displayed in the UI will be formatted based on OS locale settings?

Author: That is correct. We'll be leveraging that capability with the intention to avoid requiring the developer needing to handle localization.

    Header="Enter a percentage:"
    PlaceholderText="2/3" 
    MinValue="0"
    MaxValue="100" 
    AreLeadingZerosTrimmed="False"
    DecimalPrecision="5"
    DoesInputRound=False" />
```

## Remarks
<!-- Explanation and guidance that doesn't fit into the Examples section. -->

<!-- APIs should only throw exceptions in exceptional conditions; basically,
only when there's a bug in the caller, such as argument exception.  But if for some
reason it's necessary for a caller to catch an exception from an API, call that
out with an explanation either here or in the Examples -->

## API Notes
<!-- Option 1: Give a one or two line description of each API (type
and member), or at least the ones that aren't obvious
from their name.  These descriptions are what show up
in IntelliSense. For properties, specify the default value of the property if it
isn't the type's default (for example an int-typed property that doesn't default to zero.) -->

<!-- Option 2: Put these descriptions in the below API Details section,
with a "///" comment above the member or type. -->

## API Details
<!-- The exact API, in MIDL3 format (https://docs.microsoft.com/en-us/uwp/midl-3/) -->

```c++ 

> runtimeclass NumberBoxValueChangedEventArgs

Reviewer: For consistency with other "...ChangedEventArgs" it would be better (and more helpful) if this class included three properties: `CurrentText`, `OldValue`, and `NewValue`.
The names `Text` and `Value` on their own are ambiguous.

`CurrentText`, and `NewValue` should be the same as accessing the properties directly from the sender but are provided as a convenience to save on the cost of looking them up. `CurrentText` rather than `Text` to avoid confusion with the property and to make it clear how it compares to the other two properties.

Having access to both the old and new values can be useful when tracking changes between valid and invalid states as would be necessary as part of custom validation logic.

I don't feel it's necessary to have access to `OldText` as it may not necessarily correspond to `OldValue` unless the two were tracked in sync. Anyway it's easy to reverse engineer the old text from the old value. If there is a good reason for needing `OldText` I'd be interested it hear it--it's just not something I've ever needed or something I can imagine a case for needing.

{
    String Text;
    Double Value
};

unsealed runtimeclass NumberBox : Windows.UI.Xaml.Controls.TextBox
{
    NumberBox();
    
    Double Value;
    
    Boolean AcceptsCalculation;
    Boolean UpDownButtonsEnabled;
    Boolean HyperDragEnabled;
    Boolean HyperScrollEnabled;
    
    Double StepSize;
  
    Boolean AreLeadingZerosTrimmed;
    Double DecimalPrecision;
    Boolean DoesInputRound;
    
    Double MinValue;
    Double MaxValue;
    

    NumberBoxTemplateSettings TemplateSettings{ get; };

>     event Windows.Foundation.TypedEventHandler<NumberBox, NumberBoxValueChangedEventArgs> ValueChanged;

Reviewer: What will trigger a `ValueChanged` event?
Will `Value` be updated (and this event triggered) three times if I entered "123" in the box?
Will the event be triggered when text changes to an invalid value?
Presumably the event won't be changed when the text changes from one invalid value to another. This will mean that multiple events won't be invoked while entering a calculation until the calculation produces something valid.

I believe the above will handle all appropriate cases and allow for appropriate validation beyond an individual control until #179 is fully implemented.

Author: Great question, @mrlacey, and thanks for your patience as I continue to catch up on my inbox. I have detailed the event behaviors here in the API Notes section: https://github.com/microsoft/microsoft-ui-xaml-specs/blob/user/savoyschuler/numberbox/active/NumberBox/NumberBox.md#behavioral-components

If anything looks out of place, please let me know!

docs.microsoft.com (http://docs.microsoft.com/uwp/toolkits/winui/).
Hopefully we'll be able to copy it mostly verbatim.
So the second audience is everyone that reads there to learn how
and why to use this API. -->

## Background
<!-- Use this section to provide background context for the new API(s) 
in this spec. -->

<!-- This section and the appendix are the only sections that likely
do not get copied to docs.microsoft.com; they're just an aid to reading this spec. -->

<!-- If you're modifying an existing API, included a link here to the
existing page(s) -->

<!-- For example, this section is a place to explain why you're adding this API rather than
modifying an existing API. -->

<!-- For example, this is a place to provide a brief explanation of some dependent
area, just explanation enough to understand this new API, rather than telling
the reader "go read 100 pages of background information posted at ...". -->

## Description
<!-- Use this section to provide a brief description of the feature.
For an example, see the introduction to the PasswordBox control 
(http://docs.microsoft.com/windows/uwp/design/controls-and-patterns/password-box). -->

> Number box is a numerics only text control with support for input formatting, validation, basic calculation, and increment stepping.

Reviewer 7: I think the use cases listed are too broad. Incrementing, decrementing and calculations target a very different set of use cases than IP addresses and phone numbers.

`NumericUpDown` is a classic WinForms control, later reimplemented by many in WPF and UWP for entering and updating values. It would be useful to have one in the platform and supporting calculations would be a somewhat useful if hidden and costly feature. Having one with a `double Value` property for binding to values would be very useful.

A `NumberBox` for phone numbers, IP numbers, SSNs and other ID codes with masking and formatting support would also be useful, but it sounds like a completely different use case. There is no single numeric value it would make sense for it to expose. It doesn't make sense to perform math operations or increment/decrement values of any of these data types.

Supporting currencies is yet another story. WinRT has no currency data type as far as I know and I doubt a `double` would suffice as a reliable money type even if you also specified a currency. It would be useful to have a `CurrencyBox` or `MoneyBox` control would be useful, but it would have to be much more specialized than a `NumberBox` or `NumericUpDown` could be.

Author: Agreed - the proposal has since been significantly scoped down to no longer support masked numerical strings - only numbers and math. 

We are still investigating how heavy inline calculation will be. I have reason to think it won't be terrible considering it is only intended to support basic math (*,-,/,+), operator precedence, and parenthetical operators but we will verify. 

Does this respond fully to everything you said?

> **Important APIs:** [NumberBox class](https://docs.microsoft.com/en-us/uwp/api/microsoft.ui.xaml.controls.numberbox)

Reviewer: It's confusing that this isn't a valid link (yet)
Might be worth calling out that this is the API that is being defined in this spec. Currently this looks like a link to another, existing, class.

Author: Thanks for letting me know this was a point of confusion. I have added a note with this [commit](https://github.com/microsoft/microsoft-ui-xaml-specs/commit/16712b492a27b86416d3354e526327ed4bc8200f) to clarify this link will be generated with MSDN documentation on publication.

## Is this the right control? 

> Use a **NumberBox** control to recieve and display formatted numeric values such as currency, IP addresses, and telephone numbers. A **NumberBox** can also be enabled with basic calculator support to compute multiplication, division, addition, and subtraction across order of operations on entered equations. 

Reviewer: Can you explain what is meant by "across order of operations"? Is this any different to standard operator precedence?

Author: Great question. I intended this to communicate support for parenthetical operators in addition to standard operator precedence. Is there a way I could word this more clearly in your view? Perhaps, the following? 

```suggestion
Use a **NumberBox** control to receive and display formatted numeric values such as currency, IP addresses, and telephone numbers. A **NumberBox** can also be enabled with basic calculator support to compute multiplication, division, addition, and subtraction across parenthetical order with standard operator precedence. 
```

Author: [Committed. ](https://github.com/microsoft/microsoft-ui-xaml-specs/commit/64ac5963b9633383333eee5331a18afd5e966d27)

## Examples
<!-- Use this section to explain the features of the API, showing
example code with each description. The general format is: 
  feature explanation,
  example code
  feature explanation,
  example code
  etc.-->
  
<!-- Code samples should be in C# and/or C++/WinRT -->

<!-- As an example of this section, see the Examples section for the PasswordBox control 
(https://docs.microsoft.com/windows/uwp/design/controls-and-patterns/password-box#examples). -->

### Create a NumberBox

XAML
```XAML

> <NumberBox AutomationProperties.Name="simple NumberBox"/>

Reviewer 5: Will this automatically set the [InputScope](https://docs.microsoft.com/en-us/uwp/api/Windows.UI.Xaml.Controls.TextBox.InputScope) property to Number, or does that need to be set here?

Reviewer 3: The Number InputScope should be the default, however if **AcceptsCalculation** is set to true, whatever numeral keyboard which also supports the Plus, Asterisk, Hyphen, and Slash keys will need to be visible I would have thought

Reviewer: `Number` seems like a good default here.

The `Number` InputScope already includes the keys for basic arithmetic (on desktop at least) as defined within this spec. I don't think there's any need to automatically change the default InputScope based on other property settings.

Reviewer 9: The [`FormulaNumber`](https://docs.microsoft.com/en-us/uwp/api/Windows.UI.Xaml.Input.InputScopeNameValue) InputScope seems appropriate here

Reviewer: `FormulaNumber` includes (on the desktop at least) lots of symbols that aren't supported by this spec and also doesn't include space, while all examples (here and in the original issue) shows spaces around the operators for clarity of reading.
![image](https://user-images.githubusercontent.com/189547/58689030-e1906c80-837d-11e9-87af-fc6a64f7ba13.png)

Author: [Answered here. ](https://github.com/microsoft/microsoft-ui-xaml-specs/blob/user/savoyschuler/numberbox/active/NumberBox/NumberBox.md#behavioral-components) Number it is!

> <NumberBox AutomationProperties.Name="simple NumberBox"/>

Reviewer: Why have you specified `AutomationProperties.Name` instead of just `Name` in these examples?
Do you mean to call attention to something specific about automation support? If so, what?

Regardless, the automation property names used in this spec serve as bad examples as automation names should not include the name of the control.

Author: You're right, @mrlacey, this was a bite from the copy paste bug w/ regard to our Xaml Controls Gallery TextBox sample. I appreciate @YuliKl, our Accessibility Champion, for educating me further on automation properties and their redundancy with Headers on TextBox and its derivatives.

Author: Resolved here: https://github.com/microsoft/microsoft-ui-xaml-specs/commit/20ba7ae9254674200d8f2d363bc08949533c37b4

```

### Add placeholder text and a header

XAML
```XAML

> <TextBlock AutomationProperties.Name="simple NumberBox" Header="Enter equation:" PlaceholderText="A + B * C" />

Reviewer 5: Is this supposed to be a NumberBox?

Author: This was a copy/paste bug throughout. Thanks, Mike!

> <TextBlock AutomationProperties.Name="simple NumberBox" Header="Enter equation:" PlaceholderText="A + B * C" />

Reviewer: And in all subsequent examples, you've used `TextBox` instead of `NumberBox`

Author: [Thank you. The copy/paste bug strikes again. ](http://www.quickmeme.com/img/21/21f09c3590670b09ae768414530f1533192e0b0b8e91f1c42fb95a5ed2f0f042.jpg) Resolved with [this ](https://github.com/microsoft/microsoft-ui-xaml-specs/commit/3c5aaa4ee175abceefdc349ae02b677a2ba87b02)commit.

```

### Enable calculation support

XAML
```XAML
<TextBlock AutomationProperties.Name="NumberBox for simple calculations" 
    Header="Enter equation:" 
    PlaceholderText="A + B * C" 
    AcceptsCalculation="True" />
```

### Add formatting

```XAML

> <TextBlock AutomationProperties.Name="NumberBox for IP addresss" 

Reviewer 5: If a more general [MaskedTextBox](https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.maskedtextbox?view=netframework-4.8) control is added later, will that obsolete this NumberBox feature? Should CustomFormatString align with MaskedTextBox.Mask?

Reviewer 3: Masks could be added to the standard TextBox control, and as well as the MaskFormat, there should also be Suffix and Prefix

![image](https://user-images.githubusercontent.com/7389110/58600129-313a3f80-827b-11e9-908e-0ee3c6017ebc.png)

Reviewer: @mdtauk If your proposal is that the prefix/suffix suggestion is to be added to the TextBox, then it would be better to create a separate issue for that idea. As it is stated that the NumberBox will inherit from TextBox then that change will be picked up by this control too. 
In general it's better to create new issues for new ideas, especially when they're about different controls. Having an discussion here about that idea risks confusing the discussion that's specific to the NumberBox spec (which is what this is.) It also enables that idea/discussion to be seen as a separate thing and not overlooked among discussion elsewhere.
For the record, I can see benefits about your suggestion but have lots of questions too.

Reviewer 3: @mrlacey I was just responding to the comment prior, I have now added an at mention to make it clearer who it was directed to.

Reviewer 5: I brought this up here because I don't want to look at this control in isolation. If we have a NumberBox now, and a MaskedTextBox later, we should think now about how they're not going to collide with each other in a confusing way.

Author: Resolving for now as @mdtauk has opened an issue for adding Prefix/Suffix to TextBox here:

https://github.com/microsoft/microsoft-ui-xaml/issues/784

Reviewer 3: > Resolving for now as @mdtauk has opened an issue for adding Prefix/Suffix to TextBox here:
> 
> [microsoft/microsoft-ui-xaml#784](https://github.com/microsoft/microsoft-ui-xaml/issues/784)

FWIW Support for Masked Formatting is separate from Prefix and Suffix properties.

    Header="Enter equation:" 
    PlaceholderText="000.000.000.000"
    FormatType="IPAddress" />
```

```XAML
<TextBlock AutomationProperties.Name="NumberBox for international telephone" 

>     Header="Enter equation:" 

Reviewer: This header doesn't match the example

Author: The copy/paste bug struck a few times in this first draft. I have since removed most instances of the Header property as this will come free off TextBox and is not valuable here.

    PlaceholderText="0 (123) 456-7890"
    FormatType="InternationalTelephone" />
```

```XAML
<TextBlock AutomationProperties.Name="NumberBox for custom input" 
    Header="Enter equation:" 
    PlaceholderText="00 ft 00 in"

>     CustomFormatString="## ft ## in"

Reviewer: What is the supported formatting style?
Will it be https://docs.microsoft.com/en-us/dotnet/standard/base-types/custom-numeric-format-strings or something different?

Author: Resolving for now as, based on feedback on the proposal + internal validation, support for numerical strings is not desired by the community within the scope of this control, and thus, the CustomFormatString properties have been removed. All value formatting can now (or should be able to be) achieved via simplified APIs. If I have missed anything here, please let me know! Thanks again, @mrlacey!

>     StripLeadingZeros="False" />

Reviewer 5: This example implies that the default value it true, is that true?

Author: That's what I initially intended. Do you think NumberBox should not automatically trim leading zeros? The push on the proposal right now is for it to support purely mathematical intentions rather than numerical strings such as phone numbers or zip codes.

Reviewer 5: I see, makes sense.

```

### Add increment and decrement stepping

XAML
```XAML
<TextBlock AutomationProperties.Name="NumberBox for items in order" 
    Header="Order:" 
    Text="2"

>     StepSize="2"

Reviewer: `StepSize="2"` What is this property? Where did it come from? Why is it not documented elsewhere?

Reviewer 3: I would assume the StepSize property is the equivalent of the **StepFrequency** that is used on the Slider control.

I guess it would control how much the value will increment by when pressing the Up and Down Arrow keys or the Spin Buttons?

Reviewer: guesses and assumptions don't matter--even if they're right. Having this documented in the spec does.
The purpose of the spec is to clearly document everything that the control should do. This matters when it comes time to implement the functionality, so the developer knows what to implement. It matters during testing, so that it's possible to verify that the control correctly does everything it is supposed to. And it matters in the future as the spec is used to produce documentation so the consumers of the control know how it works.

Having an undocumented property in an example will only lead to questions and uncertainty. The purpose of reviewing the spec is to make sure there isn't any.

Where a question is asked in a review comment it is not beneficial to further discuss this in comments, unless the question is unclear and needs clarification. The correct response should be to update the document so that no-one else would ask that question upon reading the document for the first time in the future. The spec needs to standalone and not rely on people also reading comments to understand it.

Reviewer 6: > I guess it would control how much the value will increment or decrement by when pressing the Up and Down Arrow keys

Will arrow keys update the value shown inside the NumberBox? Left/Right arrows have to move focus within the NumberBox. I suppose Up/Down arrows can change the value, but that's not typical for Xaml controls. Slider is the one example I can think of.

Reviewer 7: In the NumericUpDown in WinRT XAML Toolkit they do increment/decrement the value.

Reviewer 5: Same in the Visual Studio designer.

Author: > guesses and assumptions don't matter--even if they're right. Having this documented in the spec does.
> The purpose of the spec is to clearly document everything that the control should do. This matters when it comes time to implement the functionality, so the developer knows what to implement. It matters during testing, so that it's possible to verify that the control correctly does everything it is supposed to. And it matters in the future as the spec is used to produce documentation so the consumers of the control know how it works.
> 
> Having an undocumented property in an example will only lead to questions and uncertainty. The purpose of reviewing the spec is to make sure there isn't any.
> 
> Where a question is asked in a review comment it is not beneficial to further discuss this in comments, unless the question is unclear and needs clarification. The correct response should be to update the document so that no-one else would ask that question upon reading the document for the first time in the future. The spec needs to standalone and not rely on people also reading comments to understand it.

Agreed - you have no resistance here! But because we have committed fully to open source - that includes starting our spec drafts open source as well. My normally drafting routine has been usurped to a degree by my efforts to respond to everyone here and prioritize resolutions on large and popular topics. Thanks for your patience in getting to this one. : )

@mdtauk's assessment aligned with my intention and I have since updated the property to match the name of the same property on Slider and added its description to the API Notes section which will later be turned into the MSDN API docs. I will come back by this example and add a comprehensive description as soon as I have finished corralling a majority of the API - it makes writing the examples easier when I can plan how to order them and where I can combine examples.

Author: > guesses and assumptions don't matter--even if they're right. Having this documented in the spec does.
> The purpose of the spec is to clearly document everything that the control should do. This matters when it comes time to implement the functionality, so the developer knows what to implement. It matters during testing, so that it's possible to verify that the control correctly does everything it is supposed to. And it matters in the future as the spec is used to produce documentation so the consumers of the control know how it works.
> 
> Having an undocumented property in an example will only lead to questions and uncertainty. The purpose of reviewing the spec is to make sure there isn't any.
> 
> Where a question is asked in a review comment it is not beneficial to further discuss this in comments, unless the question is unclear and needs clarification. The correct response should be to update the document so that no-one else would ask that question upon reading the document for the first time in the future. The spec needs to standalone and not rely on people also reading comments to understand it.

@mrlacey, Agreed - you have no resistance here! But because we have committed fully to open source - that includes starting our spec drafts open source as well. My normally drafting routine has been usurped to a degree by my efforts to respond to everyone here and prioritize resolutions on large and popular topics. Thanks for your patience in getting to this one. : )

@mdtauk's assessment aligned with my intention and I have since updated the property to match the name of the same property on Slider and added its description to the API Notes section which will later be turned into the MSDN API docs. I will come back by this example and add a comprehensive description as soon as I have finished corralling a majority of the API - it makes writing the examples easier when I can plan how to order them and where I can combine examples.

>     UpDownArrowsEnabled="True"

Reviewer 5: I think it would be better to be more abstract about input devices and not call out arrows in particular. E.g. one property that could enable arrows, mouse wheel, drag, etc. How will this work with a Gamepad?

Author: I [changed UpDownArrows to UpDownButtons](https://github.com/microsoft/microsoft-ui-xaml-specs/commit/d0fd39f0c7686bde60b9bd5508d1fb6b65d949ab) to clarify. Thank you for this callout.

Reviewer 3: > I [changed UpDownArrows to UpDownButtons](https://github.com/microsoft/microsoft-ui-xaml-specs/commit/d0fd39f0c7686bde60b9bd5508d1fb6b65d949ab) to clarify. Thank you for this callout.

@SavoySchuler  Fabric Web calls them **SpinButton**, I think Win32 called them **Up-Down Controls** or **Spin Controls**
![image](https://user-images.githubusercontent.com/7389110/58665150-24bff080-8328-11e9-9c06-e99b8f721a51.png)
https://docs.microsoft.com/en-us/windows/desktop/uxguide/ctrl-spin-controls

Author: Unresolving, will update this.

Author: [Updated to SpinButton.](https://github.com/microsoft/microsoft-ui-xaml-specs/commit/7ecc889b517c0b89839df29b441e2427a47cd8cf)

    UpDownDragEnabled="True" />
```

> ### Add a prefix and suffix 

Reviewer: How will prefix and suffix values be included in automation output?

Reviewer 3: Ideally I would have thought it should read out like one of these:
- 1024 (MB) -> One Thousand and Twenty Four Megabytes
- 1024 (MB) -> One Zero Two Four Megabytes

Author: Resolving for now as we have scoped out prefix/suffix for NumberBox and deferred the improvement to its parent control, TextBox.

> ### Add a prefix and suffix 

Reviewer: Will there be any built-in way to support common (I presume) prefix and suffix values that vary by culture?
As an example, I'm thinking of currency symbols which some cultures put at the front and some at the end.
If this is out of scope please can it be documented as such.

Author: You are right that there is likely a historic issue with automated localization on this front.  LTR vs RTL should be noted a point of localization automation opportunity for prefix/suffix. This should be captured on the proposal-to-be for adding this feature improvement to NumberBox's parent class, TextBox. As noted earlier, wanted to make sure one of the original idea initiators had time to add that proposal, but I will come through and add it next week if I do not see it before then. Otherwise, I will resolve this issue for this forum since prefix/suffix have been scoped out here. I'm excited to see this solved more broadly via TextBox!

XAML
```XAML
<TextBlock AutomationProperties.Name="NumberBox for IP address" 
    Header="Enter equation:" 
    PlaceholderText="000.000.000.000" 

>     Prefix="IP Address:" />

Reviewer 5: Probably need to be able to set the margin for the prefix/suffix.

Reviewer 3: I would have thought Prefix and Suffix should have templates and be stylable

Author: Resolving for this control though the discussion will be relevant and righteous for a broader "Prefix/Suffix for TextBox" proposal.

>     Prefix="IP Address:" />

Reviewer 5: Is there a scenario for both a header and a prefix? (This example doesn't look like it is one.)

Reviewer 3: Header = Enter size:
Suffix = MB
Value = 1024

Reviewer 5: But what about both a Header and a Prefix?

Author: This was copy/paste bug. Most instances of Header removed as this property will come off TextBox.

>     Prefix="IP Address:" />

Reviewer: What's the point of the prefix in this example? It seems to duplicate what could otherwise be in the header or a label.

Reviewer 3: For my visual example, I chose to use a currency symbol as an example
![image](https://user-images.githubusercontent.com/7389110/58633221-c2db9880-82df-11e9-8a83-ae515b6dccc1.png)

Author: Updated with this [commit ](https://github.com/microsoft/microsoft-ui-xaml-specs/commit/8f4f1b787152644c1c3d7d938e740a8157b60e50)to use currency as an example and removed headers from most examples as this property will come from TextBox.

Author: But yes, you're both right. @mdtauk's example is better than what was shown in the first draft and was updated to match his suggestion before being deleted due to the reason above.

>     PrefixText="IP Address:" />

Reviewer 6: In this example, it feels to me like IP Address would be a more appropriate Header than Prefix. In general, whether something is a prefix or a suffix feels very specific to a given locale, so this concept may localize very poorly. Are PrefixText/SuffixText required properties or can we ship without them?

Reviewer 3: @YuliKl Prefix and Suffix add context to the value, so suffixes like Centimetres, **cm** or Gigabytes **GB** - prefixes like Currency symbols and so on.

The header would be an instruction like **Enter partition size**, **Enter a length** or **Set your price**

Author: Resolving for now as we have scoped out prefix/suffix for NumberBox and deferred the improvement to its parent control, TextBox.

```

XAML
```XAML
<TextBlock AutomationProperties.Name="NumberBox for cents" 
    Header="Enter equation:" 

>     PlaceholderText="0.00" 

Reviewer: I'm not sure this placeholder text makes sense in the rest of this example. Unless the box is being used to capture values less than a whole cent.

Reviewer 3: This example did not make sense to me either, so for my visual example I chose to use a unit of measurement.
![image](https://user-images.githubusercontent.com/7389110/58632816-ad19a380-82de-11e9-8d22-5391d314bad3.png)

Author: Revised to your specification with [this ](https://github.com/microsoft/microsoft-ui-xaml-specs/commit/eda1ede5f672ebbefe466945cc271ee316dab6f0)commit.

Reviewer: but now the automation name doesn't correspond with the suffix.

Author: [Fixed (I think)!](https://github.com/microsoft/microsoft-ui-xaml-specs/commit/d43756e65a29c0a3158aeb1bd9a38f92e944fc1b) Will pause to start adding @mdtauk's mock up pictures now.

    Suffix="cents" />
```

### Add mininum and maximum values

XAML
```XAML
<TextBlock AutomationProperties.Name="NumberBox for items in order" 
    Header="Order:" 
    PlaceholderText="1" 
    MinValue="1"
    MaxValue="99" />
```

> ## Remarks

Reviewer: It's not explicitly called out anywhere, but I assume the control will expose the entered (or calculated) value via the `Text` property provided from the underlying TextBox control. 
- If this is the case it will be important to call this out as it would mean that a scenario where a `Quantity` property that is specified as an int in a ViewModel and is bound to a `NumberBox` control in the View would require a converter. (Or are there, currently undocumented, plans to provide conversion for scenarios like this to be built into the control?
- If this is not the case, then this needs explaining.

Author: Great question. NumberBox's main input field will be the Text property that comes off of its parent class, TextBox. We have made considerations that storing an easily accessible Double, Value, beside Text would be a convenience for developers. 

Effectively, Text and Value should always match (aside from being a conversion of each other) and when one is changed, events will be fired as it is validated before the opposite is updated. I have defined the current order and firing times of these events in API Notes > Behavioral Components > Events. 

I believe we have a more up-to-date thread on this in your review further down but please let me know if there are any absences of clarity or anything you would change!

<!-- Explanation and guidance that doesn't fit into the Examples section. -->

<!-- APIs should only throw exceptions in exceptional conditions; basically,
only when there's a bug in the caller, such as argument exception.  But if for some
reason it's necessary for a caller to catch an exception from an API, call that
out with an explanation either here or in the Examples -->

## API Notes
<!-- Option 1: Give a one or two line description of each API (type
and member), or at least the ones that aren't obvious
from their name.  These descriptions are what show up
in IntelliSense. For properties, specify the default value of the property if it
isn't the type's default (for example an int-typed property that doesn't default to zero.) -->

<!-- Option 2: Put these descriptions in the below API Details section,
with a "///" comment above the member or type. -->

## API Details
<!-- The exact API, in MIDL3 format (https://docs.microsoft.com/en-us/uwp/midl-3/) -->

```c++ 

> enum NumberBoxFormatType

Reviewer 5: Need to rationalize this with [InputScopeNameValue](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Input.InputScopeNameValue) and [CoreTextInputScope](https://docs.microsoft.com/uwp/api/Windows.UI.Text.Core.CoreTextInputScope).

Author: Resolving has support for numerical strings has been scoped out.

> enum NumberBoxFormatType

Reviewer: What are the implications of selecting different values for this enum?

Author: Resolving for now as, based on feedback on the proposal + internal validation, support for numerical strings is not desired by the community within the scope of this control, and thus, NumberBoxFormatType is vestigial.

> {

Reviewer: Is a `Custom` value needed for use as the default and in combination with `CustomFormatString`?
Or is the intention of the use of this Enum than it be nullable?

Author: Resolving for now as, based on feedback on the proposal + internal validation, support for numerical strings is not desired by the community within the scope of this control, and thus, the NumberBoxFormatType and CustomFormatString properties have been removed. You are totally correct, otherwise!

>     IPAddress,

Reviewer: Surely, need both `IPAddressv4` and `IPAddressv6`.
If not, which is this meant to represent?

Reviewer 11: It would be nice to be able to use a template rather than bake specific cases into the control, e.g. "{}.{}.{}.{}" for IPv4

Author: Resolving for now as, based on feedback on the proposal + internal validation, support for numerical strings is not desired by the community within the scope of this control. Improved TextBox masking capabilities appear to be the preferred feature improvement required for this scenario.

>     InternationalTelephone,

Reviewer: What if want to capture just a local telephone number. Will this cause an issue?
Is a `LocalTelephone` value needed too?

Author: Resolving for now as, based on feedback on the proposal + internal validation, support for numerical strings is not desired by the community within the scope of this control, and thus, FormatTypes is vestigial.

    Currency,
};

unsealed runtimeclass NumberBox : Windows.UI.Xaml.Controls.TextBox
{
    NumberBox();
    

>     String Prefix;

Reviewer: If `Prefix` and `Suffix` must be strings, then names that are more consistent with other types and properties would be `PrefixText` and `SuffixText`

Author: Agreed and updated with this [commit](https://github.com/microsoft/microsoft-ui-xaml-specs/commit/ab5324e9522d05b071754e41af27c0af6df6f34d).

    String Suffix;
    

>     Boolean AcceptsCalculation;

Reviewer: How will calculation be triggered? Automatically, on pressing the enter key, on lost focus, or some other way?
Will the change after calculation be part of the undo stack within that control, so it's possible to go back and view the original sum after the result has been displayed?

Author: @sonnemaf's initial gif shows calculation on loss of focus. I believe this is the most direct and intuitive form of interaction but I will need to engage with a designer and member of the Fluent Design System rep to validate this. Is there an alternative you prefer, @mrlacey?

Reviewer 3: @SavoySchuler  Could loss of focus AND pressing Enter, trigger the calculation?

Reviewer 11: you mean OR?

Reviewer 3: > you mean OR?

Indeed I did mean OR

Reviewer: If calculation is done on lost focus there's no way for automation/accessibility tools to report the final value without then returning to the control. This seems like an undesirable experience. 
Once focus is lost I'm also assuming that the ability to return to the original sum will also be lost from the undo stack. Maintaining the original sum in the undo stack (or not) needs to be documented in this spec. It would be nice to have if calculation is done before lost focus but if the calculation is done on lost focus then getting the original sum back is essential.

If the calculation is only done on lost focus, doesn't this also create issues for validation. For example, if calculation is done after focus is lost and it results in a value that is outside of the accepted range, what gets shown in the control? An error state? The max/min value (based on which boundary was exceeded)? - More thought is needed in this area.

Reviewer 6: Agreed with @mdtauk that Enter should trigger the calculation. That's a commonly used shortcut in the Calculator app and I believe will be intuitive for and expected by many users. Being forced to move focus to see the results of the calculation feels cumbersome and slow.

Reviewer 5: In OneNote it's triggered by "= " (equals then space), enter, and tab. And the calculation is its own frame in the Undo stack.

Reviewer 11: In Adobe XD there's not need for =, you can just write simple arithmetic directly. And it's the "formula" doesn't get saved.

Author: Loss of focus, "=", "Enter" and stepping (hyper drag, hyper scroll, up/down arrow keys, StepButton) are currently listed as the events/key presses that trigger evaluation. Does this seem comprehensive and acceptable to everyone?

The change after calculation is currently planned to be part of the undo stack.

    Boolean UpDownArrowsEnabled;

>     Boolean UpDownDragEnabled;

Reviewer: How will "UpDownDrag" work?
Will it be limited to being within the control?
How much will the value change based on the amount dragged?
What happens if this control is inside a scrollable container?
Will this override current drag behavior of selected text within the control? If so, how?

Reviewer 3: I know Xbox has the concept of a focus trap, but I don't believe this is a thing on the desktop.  Ideally with the box in focus, it would trap the drag motion.

Reviewer: @mdtauk Then that would break other functionality of a standard TextBox such as selecting text and dragging it elsewhere.
I'm trying to highlight that this is something that the spec needs to cover. 
If there are precedents elsewhere then that's great because there are lots of potential edge cases here that need to be considered, understood, and documented.

Reviewer 12: I'm assuming the `UpDownDragEnabled` actually refers to the dragging of the arrows (as you say, dragging of text doesn't make much sense). If this is the case, perhaps it should be renamed to `UpDownArrowsDragEnabled`? Dragging of the arrows would actually be pretty useful, many apps already manually implement this.

Reviewer 3: @mrlacey Microsoft Blend uses this drag mechanic with its property TextFields.  It also changes the cursor to a four way arrow to indicate it behaves differently.

It could also be a slider control built into a TextField.  If there is no way to implement the drag mechanic, we could do what Adobe does and include a dropdown/flyout with a slider.
![image](https://user-images.githubusercontent.com/7389110/58650407-7ce5fb00-8306-11e9-8bad-5bf56e5d87fe.png)

Reviewer 7: There is a way to do it with just mouse drag - see WinRT XAML Toolkit's `NumericUpDown` again
![266B14D2-4A29-4A3B-93F6-1D05C255A51D](https://user-images.githubusercontent.com/1753264/58669048-be1ee080-82ef-11e9-9b08-652c29efb405.GIF)

Reviewer 3: > There is a way to do it with just mouse drag - see WinRT XAML Toolkit's `NumericUpDown` again
> ![266B14D2-4A29-4A3B-93F6-1D05C255A51D](https://user-images.githubusercontent.com/1753264/58669048-be1ee080-82ef-11e9-9b08-652c29efb405.GIF)

@xyzzer  This is how the Blend boxes work.  But I think there may be some accessibility concerns, but thats for the devs to sort out lol

Author: I have added the [API Note](https://github.com/microsoft/microsoft-ui-xaml-specs/commit/b7786665bb18f18d02a6133abedaa8c6d9300fbd) to align with WinRT XAML Toolkit's NumericUpDown. Thanks for surfacing this, @xyzzer, it will enable a consistent experience and help us be more effective!

>     NumberBoxFormatType FormatType;

Reviewer 5: Naming guideline is "Kind" rather than "Type" (to avoid confusion with GetType()).

Author: [Updated.](https://github.com/microsoft/microsoft-ui-xaml-specs/commit/f3a32e2640b457a149f5e5e74fcd0c383dd44d77)

>     String CustomFormatString;

Reviewer: If specify a `FormatType` and a `CustomFormatString`, which takes priority?

Author: Resolving for now as, based on feedback on the proposal + internal validation, support for numerical strings is not desired by the community within the scope of this control, and thus, the NumberBoxFormatType and CustomFormatString properties have been removed.

>     Boolean StripLeadingZeros;

Reviewer 5: Sounds like a method name (begins with an imperative verb). AreLeadingZerosTrimmed?

Author: [Updated. ](https://github.com/microsoft/microsoft-ui-xaml-specs/commit/20d7f5326113aac85838dc7d7df4bb0e9cd2d95a)

>     Integer MinValue;

Reviewer: If `MinValue` and `MaxValue` are specified as integers (I'm assuming you mean `System.Int32`) this could be very limiting in a number of ways.
- It stops the specifying of numbers outside this range.
- It implies that there must be values here. (Unless these will be nullable)
- It doesn't work with floating point values (especially relevant for currency)
- It doesn't work for formatted values that aren't traditional "numbers" such as IP addresses.

Author: You're totally right, @mrlacey. To resolve the first three bullet points, these have been appropriately updated to Double. To the fourth bullet point, support for numerical strings (such as IPv4 addresses) has been scoped out. More info on the proposal.

    Integer MaxValue;

>     static Windows.UI.Xaml.DependencyProperty PrefixProperty{ get; };

Reviewer: How will the `Prefix` and `Suffix` be styled?
Alignment, spacing, foreground and background colors, and more need to be considered and the limits of customization also need to be specified.
Ideally, I'd like to see individual data templates be exposed for both these properties so they can be customized in the same way as `Header` without needing to retemplate the whole control. 
Without being able to see the intended designs for these properties it's hard to say how useful they would be by default.

Reviewer: Additionally, and for confirmation, I'm assuming that in LTR languages, the prefix is displayed on the left and the suffix on the right. For RTL languages this will be reversed.

Author: This feedback should definitely be noted on the proposal-to-be for adding this feature improvement to NumberBox's parent class, TextBox. I will resolve this issue for this forum since prefix/suffix have been scoped out here. 

I wanted to make sure one of the original idea initiators had time to add that proposal, but I will come through and add it next week if I do not see it before then. Thanks for the always excellent input, @mrlacey!

    static Windows.UI.Xaml.DependencyProperty SuffixProperty{ get; };
    

>     static Windows.UI.Xaml.DependencyProperty CalculationEnabledProperty{ get; };

Reviewer 11: This could be AcceptsCalculation just like TextBox.AcceptsReturn

Author: Great call out, @adrientetar, I have it updated!

>     static Windows.UI.Xaml.DependencyProperty UpDownArrowsEnabledProperty{ get; };

Reviewer: How will the Up and Down arrows by styled, and will there be any way to customize the styling?
A full DataTemplate might be overkill but having the ability to specify button colors is a necessity.

Reviewer: How much will the arrow buttons increment or decrement the value?
Surely this needs to be configurable, so we need a `StepSize` property too?

Author: Yes, the StepSize property has been added. As far as customization goes, great call out. Ill dig into this after we have landed on a design as we are not sure where the Up/Down buttons will go (e.g., in the control vs outside of it).

Author: I have added an open question for SpinButton styling.

Reviewer 4: @SavoySchuler About the SpinButton terminology: Perhaps because I'm a non-native English speaker, but just looking at that name, I have absolutely no clue what is meant with a *SpinButton* for this control. It also isn't intuitive to me because I only know the word *spin* in a context such as "wheel/disk spinning", however, what is spinning when looking at the NumberBox control?

Author: @Felix-Dev Thank you for mentioning this! I really appreciate the unique perspective you can bring. 

Our [Fabric UI system names this same control SpinButton](https://developer.microsoft.com/en-us/fabric#/controls/web/spinbutton) which is what I based the name off of.  Historically, names like [Spin Button](https://www.c-sharpcorner.com/uploadfile/mahesh/spin-button-in-wpf/), [Spin Control](https://docs.microsoft.com/en-us/windows/desktop/uxguide/ctrl-spin-controls), and [Spinner ](https://docs.microsoft.com/en-us/windows/desktop/windowsribbon/windowsribbon-controls-spinner)have also been used by our older frameworks as well. 

I had it named UpDownButton briefly but re-aligned for consistency across our modern and legacy platforms. Do you have some alternate suggestions that I could consider? Otherwise, perhaps I can be proactive with clear documentation examples + images for this component?

Reviewer 4: @SavoySchuler Sorry for the late reply! I have yet to think about a good alternative name.

Out of interest, what was the reason why this terminology was chosen in the first place? Perhaps that could help me relate the term "Spin Button" with the up/down buttons of the NumberBox control.

Adding images to the final documentation on docs.microsoft.com describing the different components of the UI sounds good to me! Perhaps similar to the [Navigation View UI documentation](https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/images/topnav-anatomy.png) (obviously scaled down for the NumberBox control)?

Author: @MikeHillberg do you have any history on the origin of the "Spin Button" name? To paraphrase the [Spin Controls documentation](https://docs.microsoft.com/en-us/windows/win32/uxguide/ctrl-spin-controls), my best guess is that the analogy arises from being able to increase uni-directional change while remaining stationary, as with some of these spinning objects that increase measured value per interaction from a (semi) stationary force:

[Merry-Go-Round](http://media.treehugger.com/assets/images/2013/10/merry-go-round-power.jpg.662x0_q70_crop-scale.jpg)
[Fidget Spinner](https://timedotcom.files.wordpress.com/2017/05/170503-fidget-spinner-school-ban.jpg?quality=85)

Otherwise, yes, I can be sure the visual documentation helps to make this clear. 😄

Author: @mrlacey, to answer your first posted question which I appear to have overlooked (apologies there) the following APIs have been added for styling the Spin Buttons:

    Windows.UI.Xaml.Style DecrementButtonStyle;
    Windows.UI.Xaml.Style IncrementButtonStyle;

>     static Windows.UI.Xaml.DependencyProperty UpDownDragEnabledProperty{ get; };

Reviewer: Will any UI affordance for dragging support be provided by default? If so what, and will it be customizable?

Author: That is prospective still. For this detail I will work with our designer and Fluent Design System representatives to pin down the user interaction model and whether dragging, scrolling, or arrow stepping will be supported by default or enableable. The most limiting factor in this respect will be the paradigm set by accessibility requirements and what is needed or perhaps prohibited among these options. Are any of these specifically a make or break need for you @mrlacey?

Reviewer: Nothing is essential here. That what "UpDownDragEnabled" really means or how it will be surfaced aren't clear is my biggest concern here.

Author: As noted below,I have added the [API Note](https://github.com/microsoft/microsoft-ui-xaml-specs/commit/b7786665bb18f18d02a6133abedaa8c6d9300fbd) to align hyper drage with WinRT XAML Toolkit's NumericUpDown/Blender which, I believe, has already solved this portion of our control equation. Are there any gaps you foresee with adopting this work, @mrlacey?

    
    static Windows.UI.Xaml.DependencyProperty FormatTypeProperty{ get; };
    static Windows.UI.Xaml.DependencyProperty CustomFormatStringProperty{ get; };

>     static Windows.UI.Xaml.DependencyProperty StripLeadingZerosProperty{ get; };

Reviewer: What about the ability to control the stripping of trailing zeros after a decimal place?
Would it be necessary to control this with a custom format?

Author: DecimalPrecision has been added to support this! Examples for the new APIs I will try to have added tomorrow. Thanks, @mrlacey! Do feel encouraged to suggest code edits or submit new PRs for APIs as you see fit. You're full awesome (and critical) ideas!

> unsealed runtimeclass NumberBox : Windows.UI.Xaml.Controls.TextBox

Reviewer 7: I think this class should derive from [`RangeBase`](https://docs.microsoft.com/en-us/uwp/api/Windows.UI.Xaml.Controls.Primitives.RangeBase), not `TextBox`

Reviewer 3: > I think this class should derive from [`RangeBase`](https://docs.microsoft.com/en-us/uwp/api/Windows.UI.Xaml.Controls.Primitives.RangeBase), not `TextBox`

The control has more in common with the TextBox control than it does a Slider, ScrollBar or ProgressBar control.  There is also a wish for it to automatically benefit from Prefix and Suffix properties which could be added to the TextBox.

Reviewer 7: I disagree. The main programmatic access to the control is through value bindings and user being able to interact with the value through text is more of an implementation/templating detail of the control. You can put a `TextBox` in the template and not even have to programmatically expose much of its interface. The control is mostly used for entering values and we should make it easily swappable with a slider, make sure that accessibility features provide value information etc. 

You can do it either way of course, but `RangeBase` has basically exactly the properties we need, so if we derive from `TextBox` - we'll be reinventing the wheel.

Reviewer 3: > I disagree. The main programmatic access to the control is through value bindings and user being able to interact with the value through text is more of an implementation/templating detail of the control. You can put a `TextBox` in the template and not even have to programmatically expose much of its interface. The control is mostly used for entering values and we should make it easily swappable with a slider, make sure that accessibility features provide value information etc.
> 
> You can do it either way of course, but `RangeBase` has basically exactly the properties we need, so if we derive from `TextBox` - we'll be reinventing the wheel.

So the Textbox becomes a TemplatePart.  How would that work when adding the Spin Buttons and integrating them into the TextBox itself?

Reviewer 7: Why do they have to be integrated into the `TextBox`? They can be next to it. You can also retemplate the `TextBox` and will likely need to do it either way.

Reviewer 4: Actually prefer the up/down buttons to be next to the control instead of being integrated into the TextBox.

Reviewer 3: > Why do they have to be integrated into the `TextBox`? They can be next to it. You can also retemplate the `TextBox` and will likely need to do it either way.

Making the buttons and TextBox integrated is part of the proposal, because separate buttons and a TextBox could be done with binding.

![image](https://user-images.githubusercontent.com/7389110/59051299-6f8bbc00-8884-11e9-8c14-6aa63940cdb6.png))

Author: Thanks, @xyzzer, we examined deriving from RangeBase and TextBox. As noted on the proposal and in the review summaries here, we have decided against deriving from any class in preference of containing TextBox to avoid aggregating unusable properties on this control. Your suggestions enabled us to have comprehensive discussions on this topic, so I want to be sure to thank you and credit you for our progress. 😄

    static Windows.UI.Xaml.DependencyProperty MinValueProperty{ get; };
    static Windows.UI.Xaml.DependencyProperty MaxValueProperty{ get; };
}
```

## Appendix
<!-- Anything else that you want to write down for posterity, but 
that isn't necessary to understand the purpose and usage of the API.
For example, implementation details. -->

## Appendix

> * Should we support some basic format types or only custom format strings? 

Reviewer: Only supporting custom formats makes me suspect that support for common, standard formats would be seen as an omission.

Author: As noted in my most recent update here and on the proposal, the pendulum has swayed against supporting numerical strings and thus formatting support for numbers has been simplified to handful of APIs.

docs.microsoft.com (http://docs.microsoft.com/uwp/toolkits/winui/).
Hopefully we'll be able to copy it mostly verbatim.
So the second audience is everyone that reads there to learn how
and why to use this API. -->

## Background
<!-- Use this section to provide background context for the new API(s) 
in this spec. -->

<!-- This section and the appendix are the only sections that likely
do not get copied to docs.microsoft.com; they're just an aid to reading this spec. -->

<!-- If you're modifying an existing API, included a link here to the
existing page(s) -->

<!-- For example, this section is a place to explain why you're adding this API rather than
modifying an existing API. -->

<!-- For example, this is a place to provide a brief explanation of some dependent
area, just explanation enough to understand this new API, rather than telling
the reader "go read 100 pages of background information posted at ...". -->

## Description
<!-- Use this section to provide a brief description of the feature.
For an example, see the introduction to the PasswordBox control 
(http://docs.microsoft.com/windows/uwp/design/controls-and-patterns/password-box). -->

> Number box is a numerics only text control with support for validation, increment stepping, and computing inline calculations.

Reviewer 13: Will the box understand hex values? Can I do "AA + 1"? It's ok to say no, but we should be crisp about what it supports and does not.

Reviewer 3: Is this something that could be a property?
So the developer can provide a valid range of chars, or maybe an enum with known alphanumeric representations of numbers?

- **ValueType.Hexadecimal**
- **ValueType.Binary**

Author: _It could be._ 

I will take this topic over to the proposal to see if there is developer/app-needed justification for it.

**Important APIs:** [NumberBox class](https://docs.microsoft.com/en-us/uwp/api/microsoft.ui.xaml.controls.numberbox) [TODO - generated with MSDN documentation on publication.]

## Is this the right control? 

> Use a **NumberBox** to capture and display mathematical input. A **NumberBox** can also be enabled with basic calculator support to compute multiplication, division, addition, and subtraction across parenthetical order with standard operator precedence.

Reviewer 13: Have we validated that these operators will be sufficient? What about exponents? Percentage? I think we should support anything that the Cortana numberbox calculator can support.

Reviewer 3: As a graphics application example...
If you had a value of **256 px**, how would you enter the operator to find the 25% value of that original value?

### 256 25%

Or do we mentally figure out that 25% is dividing the value by 4?

### 256 / 4

Reviewer 13: Good question. I don't have a strong opinion on percentage specifically; my meta-feedback is that NumberBox will have a list of supported operators, and I just wanted to ensure that we were being thoughtful about which operators we support. (The percent sign can also be modulo, so that adds a potential level of complexity to this specific operator.)

Author: Initiated required operators conversation up-level on the feature proposal. Closing here to maintain a single source thread. Thanks @stmoy and @mdtauk for starting this!

from their name.  These descriptions are what show up
in IntelliSense. For properties, specify the default value of the property if it
isn't the type's default (for example an int-typed property that doesn't default to zero.) -->

<!-- Option 2: Put these descriptions in the below API Details section,
with a "///" comment above the member or type. -->

## API Notes

### Notable Properties  

| Name | Description |
|:-:|:--|
| StepFrequency | Gets or sets the value part of a value range that steps should be created for. |

## API Details
<!-- The exact API, in MIDL3 format (https://docs.microsoft.com/en-us/uwp/midl-3/) -->

```c++ 
runtimeclass NumberBoxValueChangedEventArgs
{
    String Text;
    Double Value
};

runtimeclass NumberBoxTextChangedEventArgs
{
    String Text;
    Double Value
};

unsealed runtimeclass NumberBox : Windows.UI.Xaml.Controls.TextBox
{
    NumberBox();
    
    Double Value;
    Boolean AcceptsCalculation;
    
    Boolean UpDownButtonsEnabled;
    Boolean HyperDragEnabled;
    Boolean HyperScrollEnabled;
    Double StepFrequency;
  
    Boolean AreLeadingZerosTrimmed;
    Double DecimalPrecision;
    Boolean DoesInputRound;
    

>     Double MinValue;

Reviewer 13: What happens to the up/down buttons when the min/max is hit. Do they disable? Consider that the buttons are enabled, the step size is 1, the minimum is 0, and the current value is 2. I click Down-->Down, so the value is currently 0. Does the down button get disabled?

Reviewer: Disabling is nice but not essential IMHO

Author: @mrlacey provided a prototype/analysis earlier and raised a good point that stepping beyond a min or max should coerce the value to the min or max, respectively. E.g., user enters value of 0.5, StepFrequency="0.2", MinValue="0.0", MaxValue="1.0", if the user were to use hyper scrolling || hyper drag ||  UpDownButtons to step, updates would stop at 0.1 or 0.9 rather than at the bounds. 

It would be great to see the buttons disabled when the MinValue/MaxValue are hit, but as @mrlacey points out, it has not been an essential ask.

Author: Added to API Notes > Behavioral Components > SpinButton.

    event Windows.Foundation.TypedEventHandler<NumberBox, NumberBoxValueChangedEventArgs> ValueChanged;
    event Windows.Foundation.TypedEventHandler<NumberBox, NumberBoxTextChangedEventArgs> TextChanged;

    static Windows.UI.Xaml.DependencyProperty ValueProperty{ get; };
    static Windows.UI.Xaml.DependencyProperty AcceptsCalculationProperty{ get; };
    
    static Windows.UI.Xaml.DependencyProperty UpDownButtonsEnabledProperty{ get; };
    static Windows.UI.Xaml.DependencyProperty HyperDragEnabledProperty{ get; };
    static Windows.UI.Xaml.DependencyProperty HyperScrollEnabledProperty{ get; };   
    static Windows.UI.Xaml.DependencyProperty StepFrequencyProperty{ get; };
    
    static Windows.UI.Xaml.DependencyProperty AreLeadingZerosTrimmedProperty{ get; };
    static Windows.UI.Xaml.DependencyProperty DecimalPrecisionProperty{ get; };
    static Windows.UI.Xaml.DependencyProperty DoesInputRoundProperty{ get; };
    
    static Windows.UI.Xaml.DependencyProperty MinValueProperty{ get; };
    static Windows.UI.Xaml.DependencyProperty MaxValueProperty{ get; };
    
    static Windows.UI.Xaml.DependencyProperty IsInvalidInputOverwrittenProperty{ get; };
}
```

> ## Appendix

Reviewer 13: The appendix is probably an appropriate place to discuss keyboard/gamepad behavior. Of note, I'm curious how the UP/DOWN buttons work, as well as the ESC key. I believe ENTER/TAB is discussed elsewhere, but it'd be worthwhile to condense that info into a single place.

Author: Agreed. I believe @Austin-Lamb has been advocating that those sections be part of the standard, required template and I intend to include it shortly.

Will leave this open until resolved.

Reviewer 3: My gut tells me **Esc**ape should cancel a change in the value that was already there.  It does not in Photoshop, with annoys me all the time I wish to cancel a change I had already started making.

Reviewer: **Esc**ape already has window level functionality at some levels such as cancelling/closing a dialog.
If the NumberBox was inside such a dialog it could give the impression that the whole dialog wasn't working as expected.
Undo (ctrl+Z) allows reverting changes that have been made.

Reviewer 13: > Escape already has window level functionality at some levels such as cancelling/closing a dialog.

So does ENTER. We have some precedent that ESC will "reset" the field - the biggest one that comes to mind is ComboBox. 

I think ENTER and ESC go hand-in-hand in that if NumberBox responds to one, it should respond to both.

Reviewer 6: > I think ENTER and ESC go hand-in-hand in that if NumberBox responds to one, it should respond to both.

I disagree, I think Enter and Escape are distinct concepts. Escaping out of a TextBox or NumberBox doesn't conceptually make sense. In ComboBox, the Esc key behavior varies depending on whether the ComboBox's flyout is open or closed. On a closed ComboBox, Escape doesn't do anything.

Reviewer 3: You would be **Esc**aping from the calculation that is about to be made in the textbox, and removing focus from it.  **Enter** would be submitting the calculation and accepting the result it gives as the new Value for the control.

Reviewer 6: Escape and undo aren't exactly the same concept. They're related, more closely in my opinion than Escape and Enter, but still distinct. We tend to equate escape with back, at least in flyout scenarios. Matt had mentioned Ctrl+Z for undo (notably, not the same as **Esc**ape), but even that, to me, feels outside the scope of NumberBox. I don't believe that the default TextBox, or any other Xaml control, has the concept of history/clipboard built in.

Reviewer 3: > Escape and undo aren't exactly the same concept. They're related, more closely in my opinion than Escape and Enter, but still distinct. We tend to equate escape with back, at least in flyout scenarios. Matt had mentioned Ctrl+Z for undo (notably, not the same as **Esc**ape), but even that, to me, feels outside the scope of NumberBox. I don't believe that the default TextBox, or any other Xaml control, has the concept of history/clipboard built in.

Just to clarify, I am not saying Escape would undo the last calculation.

If you enter in some text like: **128 / 2**, but before you actually press Enter to actually do the calculation, and replace the Value and Text with the calculated value - pressing Escape would remove the text you have entered, will not do the calculation, and will replace the Text with the Value property, putting the value back the way it was before you started typing.

Reviewer 6: I'll leave it to @SavoySchuler to spec, but to me this doesn't quite feel like the right use of Escape, and I think that our [Esc key documentation](https://docs.microsoft.com/en-us/windows/uwp/design/input/keyboard-interactions#enter-key) supports this point of view: *"The Esc key lets a user cancel transient UI (along with any ongoing actions in that UI)."*

Reviewer 3: I believe the Calculation you type into the Text property is Transient until it is confirmed by pressing Enter, or clicking off the NumberBox after entering a valid calculation string.  There will possibly be a preview value displayed whilst the NumberBox still has focus, but having the ability to back out of the calculation, would be useful.

Reviewer 9: And in general, single-line text boxes don't eat the Enter key. That's reserved for form submission or dialog confirmation and we should be careful about this control getting in the way of that.

Author: Thanks for building up a pro/con case for each! It appears that that behaviors that may be desirable from the user perspective may not align with broad expectations our ecosystem has created through legacy. I will have both Escape and Enter reviewed today by the broader team shortly.

Author: Resolving conversation: loss of focus, "Enter", and stepping will trigger evaluation, as noted below in review summary and in spec.

<!-- Anything else that you want to write down for posterity, but 
that isn't necessary to understand the purpose and usage of the API.
For example, implementation details. -->

### Behavioral Components

| Property | Notes |
|:---:|:---|

> | Value/Text Changed | * When Text is changed via codebehind or by user input ("Enter" is pressed or the NumberBox loses focus) the NumberBox will be prompted to update its Text and Value properties. A TextChanged event will be raised that will exposes the new Text and the old Value and allows the developer to intercept the and manipulate these properties (such as for manually handling validation error). Value will then be updated to the Text input converted to a Double. <br><br> * Conversely, code behind updates to Value will raise a ValueChanged event that will exposes the new Value and the old Text and allows the developer to intercept the and manipulate these properties (such as for manually handling validation error). Text will then be updated to the Value input converted to a String.    |

Reviewer 13: Scenarios corresponding to error states would be great to explicitly call out and have a point of view on. I'm still not sure if/how we handle error states - is it an exception? A piece of UI? How does error states affect the "Value" property? How do they work with the input validation stuff mentioned here? 

Specific error states that are on my mind are:
* Div by zero
* Min or Max reached/exceeded
* Incorrect text entered
* Unsupported operator

Reviewer 3: For Min or Max values, rather than returning the value to the previous one, there may be a case to clamp values above to the Max value, or values below to the Min value.

Reviewer 13: Probably. I can see a bunch of potential possibilities including: 
* clamp to the min/max
* return to last known good value
* wrap around

I think we need to figure out a default behavior and then let the dev have the capability to do the others.

## Open Questions

> * For touch/virtual keyboards, can we intelligently swap to the numerical input keyboard?

Reviewer: I'm not sure I understand this question. Do you mean, 'Can the InputScope always be set to "Number?"'

Author: @robloo brought up the idea of visually limiting on-screen keyboard to numerical input only when a NumberBox is the target of focus for the clarity and convenience of end users. I will need to familiarize myself further with input scoping/on-screen keyboard to see if this functionality is something that on-screen keyboard is already designed to support.

Reviewer 3: It would be nice if the On-Screen Keyboard could include a "Calculator" layout that could be useful.  But that is not in the control of WinUI 3.0 ...

https://docs.microsoft.com/en-gb/windows/uwp/design/input/use-input-scope-to-change-the-touch-keyboard

Number seems to be the only option that makes sense

Author: Currently this kind of configuration for on-screen keyboard is not available as this keyboard configuration is redundant to calculator scenarios (which show this scope limited input display on-screen already. Closing for now but can revisit for V2.

## Open Questions

* Consenus on Validation behavior. [See API Notes > Behavioral Components]

* Is there value in creating a preview for calculation results? @mdtauk created a few example visualizations: 

![NumberBox with a tool tip above to show a preview of the calculation results](https://user-images.githubusercontent.com/16964652/58919441-fbfe7900-86e2-11e9-8d2b-dd4dadfa74c5.png)

![NumberBox with a calculation in progress and highlight text previewing the calculation results](https://user-images.githubusercontent.com/7389110/58920708-5b807700-872b-11e9-9924-21a7b7d37e68.png)

* InputScope: Number or FormulaNumber?

*  Should we rely on the Xaml ControlTemplate for creating a NumberBox with disjoint UpDownButtons (i.e., this is not common enough to justify supporting out of the box) or should we include a property for setting whether the UpDownButtons appear appear contiguous vs. disjoint?

> ![one diagram showing UpDownButtons side-by-side and another diagram showing the down button to the left of the NumberBox and the up button to the right of the NumberBox](https://user-images.githubusercontent.com/16964652/58919787-3ae0fe80-86e4-11e9-8136-4c5b204825b0.png)

Author: I will leave this unresolved to keep the designs surfaced for V2. For V1, scope has been set at an inline SpinnerButton added via enum so that the control remains more easily extensible for adding this display modes.  

Beautiful designs as always, @mdtauk!

> |Decimal Precision | * Positive DecimalPrecision values truncate post-decimal values. E.g., DecimalPrecision="5", DoesInputRound="False", input is 6.1234567, Text="6.12345" on evalutation. E.g., DecimalPrecision="5", input is 6.123, Text="6.12300" on evalutation. <br><br> * Negative DecimalPrecision values truncate pre-decimal values. E.g., DecimalPrecision="-3", DoesInputRound="False", input is 54321, Text="54000" on evalutation. E.g., DecimalPrecision="-3", DoesInputRound="True", input is 54321, Text="55000" on evalutation. <br><br> * "0" gets auto-filled on leading decimal. E.g., input is .4, Text="0.4" on evaluation. 

Reviewer 9: How does this proposal line up to the properties on [DecimalFormatter](https://docs.microsoft.com/en-us/uwp/api/windows.globalization.numberformatting.decimalformatter)?

Author: Removed. This was vestigial after adding NumberBoxNumberRounder and associated properties.
